runtime/amd64.S

```diff
@@ -916,70 +916,73 @@ FUNCTION(G(caml_raise_exn))
-#if !defined(WITH_THREAD_SANITIZER)
         testq   $1, Caml_state(backtrace_active)
         jne   LBL(116)
         RESTORE_EXN_HANDLER_OCAML
         ret
-#endif
 LBL(116):
         movq    $0, Caml_state(backtrace_pos)
 LBL(117):
         movq    %rsp, %r13        /* Save OCaml stack pointer */
         movq    %rax, %r12        /* Save exception bucket */
         movq    Caml_state(c_stack), %rsp
-#if defined(WITH_THREAD_SANITIZER)
-        testq   $1, Caml_state(backtrace_active)
-        je    LBL(118)
-#endif
         movq    %rax, C_ARG_1                    /* arg 1: exception bucket */
         movq    STACK_RETADDR(%r13), C_ARG_2     /* arg 2: pc of raise */
         leaq    STACK_ARG_1(%r13), C_ARG_3       /* arg 3: sp at raise */
         movq    Caml_state(exn_handler), C_ARG_4 /* arg 4: sp of handler */
         C_call  (GCALL(caml_stash_backtrace))
-#if defined(WITH_THREAD_SANITIZER)
-LBL(118):
-    /* Signal to TSan all stack frames exited by the exception. No need to save
-       any registers here. */
-        movq    STACK_RETADDR(%r13), C_ARG_1     /* arg 1: pc of raise */
-        leaq    STACK_ARG_1(%r13), C_ARG_2       /* arg 2: sp at raise */
-        movq    Caml_state(exn_handler), C_ARG_3 /* arg 3: sp of handler */
-        C_call  (GCALL(caml_tsan_exit_on_raise))
-#endif
         movq    %r12, %rax        /* Recover exception bucket */
         RESTORE_EXN_HANDLER_OCAML
         ret
 CFI_ENDPROC
 ENDFUNCTION(G(caml_raise_exn))
 
-/* Exception raising routine called in case of Raise_notrace when TSan
-   is enabled. Does not store a backtrace but still signals function
-   exits to TSan. */
 #if defined(WITH_THREAD_SANITIZER)
-FUNCTION(G(caml_tsan_raise_notrace_exn))
+/* When TSan support is enabled, this routine should be called just before
+   raising an exception. It calls __tsan_func_exit for every OCaml frame about
+   to be exited due to the exception.
+   Takes no arguments, clobbers C_ARG_1, C_ARG_2, C_ARG_3 and potentially all
+   caller-saved registers of the C calling convention. */
+FUNCTION(G(caml_tsan_exit_on_raise_asm))
 CFI_STARTPROC
         ENTER_FUNCTION
-        movq    %rsp, %r13        /* Save OCaml stack pointer */
-        movq    %rax, %r12        /* Save exception bucket */
-        movq    Caml_state(c_stack), %rsp
-        jmp   LBL(118)
+        movq    STACK_RETADDR(%rsp), C_ARG_1   /* arg 1: pc of raise */
+        leaq    STACK_ARG_1(%rsp), C_ARG_2  /* arg 2: sp at raise */
+        movq    Caml_state(exn_handler), C_ARG_3 /* arg 3: sp of handler */
+        SWITCH_OCAML_TO_C
+        C_call  (GCALL(caml_tsan_exit_on_raise))
+        SWITCH_C_TO_OCAML
+        LEAVE_FUNCTION
+        ret
+CFI_ENDPROC
+ENDFUNCTION(G(caml_tsan_exit_on_raise_asm))
+
+/* Like caml_tsan_exit_on_raise_asm, this routine calls __tsan_func_exit for
+   every OCaml frame about to be exited due to the exception. In addition, it
+   proceeds to raising the exception. */
+FUNCTION(G(caml_tsan_exit_on_raise_c_asm))
```

Comment:
There is a mismatch between the naming and the semantics of this function -- the name does not at all suggest that it "continues with raising". Could we get rid of the "continues with raising" aspect, by calling this function from `caml_raise_exception` instead of jumping to it?

Reply:
Calling it raises the additional difficulty that it pushes a return address onto the OCaml stack, which must be taken into account when calling `caml_tsan_exit_on_raise`. While this could be done, another possibility is to simply inline this routine at its (only) call site. I did this in 034df9f. Alternatively, we could also find a better name for the routine.

Reply:
Yes, I like the inlined version better, thanks!